XamarinCommunityToolkit/Helpers/WeakEventManagerService.shared.cs

```diff
@@ -0,0 +1,178 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using Xamarin.CommunityToolkit.Exceptions;
+
+// Inspired by AsyncAwaitBestPractices.WeakEventManagerService: https://github.com/brminnick/AsyncAwaitBestPractices
+namespace Xamarin.CommunityToolkit.Helpers
+{
+    static class EventManagerService
+    {
+        internal static void AddEventHandler(in string eventName, in object handlerTarget, in MethodInfo methodInfo, in Dictionary<string, List<Subscription>> eventHandlers)
+        {
+            var doesContainSubscriptions = eventHandlers.TryGetValue(eventName, out var targets);
+            if (!doesContainSubscriptions || targets is null)
+            {
+                targets = new List<Subscription>();
+                eventHandlers.Add(eventName, targets);
+            }
+
+            if (handlerTarget is null)
+                targets.Add(new Subscription(null, methodInfo));
+            else
+                targets.Add(new Subscription(new WeakReference(handlerTarget), methodInfo));
+        }
+
+        internal static void RemoveEventHandler(in string eventName, in object handlerTarget, in MemberInfo methodInfo, in Dictionary<string, List<Subscription>> eventHandlers)
+        {
+            var doesContainSubscriptions = eventHandlers.TryGetValue(eventName, out var subscriptions);
+            if (!doesContainSubscriptions || subscriptions is null)
+                return;
+
+            for (var n = subscriptions.Count; n > 0; n--)
+            {
+                var current = subscriptions[n - 1];
+
+                if (current.Subscriber?.Target != handlerTarget
+                    || current.Handler.Name != methodInfo?.Name)
+                {
+                    continue;
+                }
+
+                subscriptions.Remove(current);
+                break;
+            }
+        }
+
+        internal static void HandleEvent(in string eventName, in object sender, in object eventArgs, in Dictionary<string, List<Subscription>> eventHandlers)
+        {
+            AddRemoveEvents(eventName, eventHandlers, out var toRaise);
+
+            for (var i = 0; i < toRaise.Count; i++)
```

Comment:
This way we will have a perf boost.
```suggestion
            var toRaiseCount = toRaise.Count;
            for (var i = 0; i < toRaiseCount; i++)
```